pkg/collector/corechecks/ebpf/probe/ebpfcheck/probe.go

```diff
@@ -523,3 +537,313 @@ func ringBufferMemoryUsage(mapStats *model.EBPFMapStats, info *ebpf.MapInfo, k *
+type entryCountBuffers struct {
+	// Buffer to store the keys returned from the batch
+	keys []byte
+
+	// Buffer to store the values from the batch
+	values []byte
+
+	// A map that stores the hashes of the keys seen in the first batch of a batch lookup
+	// This is only used when the buffer limits do not allow us to get all the entries in a single batch
+	// and we need to iterate. In that case, we need to check against the keys we got in the first batch
+	// to see if we got restarted.
+	firstBatchKeys inplaceSet
+
+	// Buffer for the cursor indicating the next key to get
+	cursor []byte
+
+	// Track the maximum size of each buffer type, to avoid multiple reallocations
+	// each time we call the function. This way, after buffers are reset, we allocate
+	// directly the maximum size we will need and save on allocations
+	maxKeysSize   uint32
+	maxValuesSize uint32
+	maxCursorSize uint32
+
+	// size limits, originating from configuration
+	keysBufferSizeLimit   uint32
+	valuesBufferSizeLimit uint32
+}
+
+// growBufferWithLimit creates or grows the given buffer with a configured limit.
+// Returns the new buffer, the length allocated and the boolean indicating if the buffer
+// was allocated the desired size
+func growBufferWithLimit(buffer []byte, newSize uint32, limit uint32) ([]byte, uint32, bool) {
+	if newSize <= uint32(len(buffer)) {
+		return buffer, uint32(len(buffer)), true
+	}
+	if limit > 0 && newSize > limit && len(buffer) < int(limit) {
+		return make([]byte, limit), limit, false
+	}
+	return make([]byte, newSize), newSize, true
+}
+
+func (e *entryCountBuffers) tryEnsureSizeForFullBatch(referenceMap *ebpf.Map) bool {
+	maxSize := referenceMap.MaxEntries()
+	var keysAllocatedAsNeeded, valuesAllocatedAsNeeded bool
+	e.keys, e.maxKeysSize, keysAllocatedAsNeeded = growBufferWithLimit(e.keys, max(e.maxKeysSize, referenceMap.KeySize()*maxSize), e.keysBufferSizeLimit)
+	e.values, e.maxValuesSize, valuesAllocatedAsNeeded = growBufferWithLimit(e.values, max(e.maxKeysSize, referenceMap.ValueSize()*maxSize), e.valuesBufferSizeLimit)
+
+	e.ensureSizeCursor(referenceMap)
+
+	return keysAllocatedAsNeeded || valuesAllocatedAsNeeded
+}
+
+func (e *entryCountBuffers) prepareFirstBatchKeys(referenceMap *ebpf.Map) {
+	e.firstBatchKeys.keySize = int(referenceMap.KeySize())
+	e.firstBatchKeys.prepare(len(e.keys) / int(referenceMap.KeySize()))
+	// Maps grow automatically and do not shrink, so it does not make sense
+	// to reallocate them if we already have a map. However, we do want to clear it
+	// so that we don't keep old keys from previous iterations
+	e.firstBatchKeys.clear()
+}
+
+func (e *entryCountBuffers) ensureSizeCursor(referenceMap *ebpf.Map) {
+	e.cursor, e.maxCursorSize, _ = growBufferWithLimit(e.cursor, max(e.maxCursorSize, referenceMap.KeySize()), 0) // No limit with cursors, they are always small
+}
+
+// resetBuffers resets the buffers to nil, so that they can be garbage collected
+func (e *entryCountBuffers) resetBuffers() {
```

Comment:
Is there a benefit to resetting the buffers? The check runs every 15 seconds, so we would just allocate again. WDYT?

Reply:
In the load tests the average memory usage was less with the reset than without it (I had [the results saved](https://private-user-images.githubusercontent.com/1334568/299635495-5d9795f3-2cba-4055-a13d-5ae14e3c6555.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MDY2MDE0MzIsIm5iZiI6MTcwNjYwMTEzMiwicGF0aCI6Ii8xMzM0NTY4LzI5OTYzNTQ5NS01ZDk3OTVmMy0yY2JhLTQwNTUtYTEzZC01YWUxNGUzYzY1NTUucG5nP1gtQW16LUFsZ29yaXRobT1BV1M0LUhNQUMtU0hBMjU2JlgtQW16LUNyZWRlbnRpYWw9QUtJQVZDT0RZTFNBNTNQUUs0WkElMkYyMDI0MDEzMCUyRnVzLWVhc3QtMSUyRnMzJTJGYXdzNF9yZXF1ZXN0JlgtQW16LURhdGU9MjAyNDAxMzBUMDc1MjEyWiZYLUFtei1FeHBpcmVzPTMwMCZYLUFtei1TaWduYXR1cmU9NWNjMGNhODFhMTYwZDVhNzVkYjA0NmNkNjAyMTNiZTQxOTUzNmE0MzEzMzE4YmNmZWQzYWMxOThlNDA1OTZjYiZYLUFtei1TaWduZWRIZWFkZXJzPWhvc3QmYWN0b3JfaWQ9MCZrZXlfaWQ9MCZyZXBvX2lkPTAifQ.Hwo8p3f1UL29mDgfZS7kkHv2aChQmutMAqTlrB0SExk)), some 30% less. Of course it's more of an "artifact" because we're taking the average, but I think that if we can free the buffers during the part of those 15 seconds we're not using them it's worth it, specially because when we want them again we can just reallocate directly the buffer with the size we need, we aren't calculating anything again.

I think that the map does need a reset in any case, because it doesn't shrink and we could run the risk of the map growing and growing indefinitely. I've seen some other instances in the agent code where the maps are reset because of a similar issue. However, I'm not 100% confident about the actual chances of that happening here.